ogcore/demographics.py

```diff
@@ -719,32 +816,91 @@ def get_pop_objs(
             imm_rates_orig,
             np.tile(
                 imm_rates_orig[-1, :].reshape(1, E + S),
-                (T - imm_rates_orig.shape[0], 1),
+                (T + S - imm_rates_orig.shape[0], 1),
             ),
         ),
         axis=0,
     )
     # If the population distribution was given, check it for consistency
     # with the fertility, mortality, and immigration rates
-    if pop_dist is not None:
-        len_pop_dist = pop_dist.shape[0]
-        pop_counter_2D = np.zeros((len_pop_dist, E + S))
-        # set initial population distribution in the counterfactual to
-        # the first year of the user provided distribution
-        pop_counter_2D[0, :] = pop_dist[0, :]
-        for t in range(1, len_pop_dist):
-            # find newborns next period
-            newborns = np.dot(fert_rates[t - 1, :], pop_counter_2D[t - 1, :])
-
-            pop_counter_2D[t, 0] = (
-                1 - infmort_rates[t - 1]
-            ) * newborns + imm_rates[t - 1, 0] * pop_counter_2D[t - 1, 0]
-            pop_counter_2D[t, 1:] = (
-                pop_counter_2D[t - 1, :-1] * (1 - mort_rates[t - 1, :-1])
-                + pop_counter_2D[t - 1, 1:] * imm_rates_orig[t - 1, 1:]
+    # if pop_dist is not None and not infer_pop:
+    # len_pop_dist = pop_dist.shape[0]
+    # pop_counter_2D = np.zeros((len_pop_dist, E + S))
+    len_pop_dist = pop_2D.shape[0]
+    pop_counter_2D = np.zeros((len_pop_dist, E + S))
+    # set initial population distribution in the counterfactual to
+    # the first year of the user provided distribution
+    # pop_counter_2D[0, :] = pop_dist[0, :]
+    pop_counter_2D[0, :] = pop_2D[0, :]
+    for t in range(1, len_pop_dist):
+        # find newborns next period
+        # newborns = np.dot(fert_rates[t - 1, :], pop_counter_2D[t - 1, :])
+
+        # pop_counter_2D[t, 0] = (
+        #     1 - infmort_rates[t - 1]
+        # ) * newborns + imm_rates[t - 1, 0] * pop_counter_2D[t - 1, 0]
+        # pop_counter_2D[t, 1:] = (
+        #     pop_counter_2D[t - 1, :-1] * (1 - mort_rates[t - 1, :-1])
+        #     + pop_counter_2D[t - 1, 1:] * imm_rates_orig[t - 1, 1:]
+        # )
+        newborns = np.dot(fert_rates[t - 1, :], pop_counter_2D[t - 1, :])
+
+        pop_counter_2D[t, 0] = (
+            1 - infmort_rates[t - 1]
+        ) * newborns + imm_rates_orig[t - 1, 0] * pop_counter_2D[t - 1, 0]
+        pop_counter_2D[t, 1:] = (
+            pop_counter_2D[t - 1, :-1] * (1 - mort_rates[t - 1, :-1])
+            + pop_counter_2D[t - 1, 1:] * imm_rates_orig[t - 1, 1:]
+        )
+    # Check that counterfactual pop dist is close to pop dist given
+    # assert np.allclose(pop_counter_2D, pop_dist)
+    assert np.allclose(pop_counter_2D, pop_2D)
+
+    """"
+    CHANGE - in OG-Core, we are implicitLy assuming pre-TP rates of mortality,
+    fertility, and immigration are the same as the period 0 rates.
+
+    So let's just infer the pre-pop_dist from those.
+    """
+    pop1 = pop_2D[0, :]
```

Comment:
@rickecon This line through 882 make the fix that solves the resource constraint error I was running into from this demographics module.

In `TPI.py` we make some implicit assumptions that the immigration rate in the pre-TPI period is equal to the the rate in the first period (see [here](https://github.com/PSLmodels/OG-Core/blob/1331b6a61a2ab52ac140a59eb76a6e9a9365613f/ogcore/aggregates.py#L136)) and that the mortality rates in the pre-TPI period is equal to the mortality rates in the first period (see [here](https://github.com/PSLmodels/OG-Core/blob/1331b6a61a2ab52ac140a59eb76a6e9a9365613f/ogcore/aggregates.py#L193)).  This means that when we are using time varying mortality and fertility, we need to be very careful that everything lines up from not just the first period of TPI onward, but also from period -1 to the first model period.  

The "fix" I'm making here is one option.  It takes the assumptions embedded in `TPI.py` (that the period -1 mortality, fertility, and immigration rates are the same as in period 0) and uses that to solve for the period -1 population distribution given the population distribution in period 0.  With this, the resource constraint issue when solving TPI is resolved.

The drawback to this approach is that the period -1 population is known from the data, but we don't use it.  However, if we were to use the known period -1 population distribution and for things to all be consistent, we would need to pass the period -1 mortality, fertility, and immigration rates into the model.  That means adding three new parameters to the dictionary returned from `get_pop_objs` and in the the `Specifications` object passed around OG-Core  *or* having the same `omega`, `imm_rates`, `rho`, objects as we do now, but adding another period at the beginning (so they become length T+S+1).  However, doing that would make the indexing of those arrays different from all others in the model and therefore confusing throughout the OG-Core code (i.e., `imm_rates[0, :]` would represent immigration rates in period -1, while `Y[0]` would represent GDP in period 0.)

I opted for the approach taken here because it keeps the rest of the code base more concise and clear and the approximation error is probably not very large.  But I'm happy to chat about doing things differently.

Reply:
@jdebacker. Yes. I like this approach.